# **TikTok Project**



The team is still in the early stages of the project. I have received notice that TikTok's leadership team has approved the project proposal. To gain clear insights to prepare for a claims classification model, TikTok's provided data must be examined to begin the process of exploratory data analysis (EDA).

In this notebook, the data will be analyzed, identify variables, and I will ensure the dataset is ready for analysis. Also, I will look for answers to some questions.

# ** Inspecting and analyzing data**

In this task, I will examine data provided and prepare it for analysis.
<br/>

**The purpose** of this project is to investigate and understand the data provided. This task will:

1.   Acquaint me with the data

2.   Compile summary information about the data

3.   Begin the process of EDA and reveal insights contained in the data

4.   Prepare me for more in-depth EDA, hypothesis testing, and statistical analysis

**The goal** is to construct a dataframe in Python, perform a cursory inspection of the provided dataset, and inform TikTok data team members of my findings.
<br/>
*This task has three parts:*

**Part 1:** Understand the situation
* How can I best prepare to understand and organize the provided TikTok information?

**Part 2:** Understand the data

* Creating a pandas dataframe for data learning and future exploratory data analysis (EDA) and statistical activities

* Compiling summary information about the data to inform the next steps

**Part 3:** Understand the variables

* Use insights from my examination of the summary data to guide deeper investigation into variables

<br/>

To complete the activity, follow the instructions and answer the questions below. Then, you will us your responses to these questions and the questions included in the Course 2 PACE Strategy Document to create an executive summary.

After completing this task, I will complete an Executive Summary using the questions listed on the PACE Strategy Document.

# **Identify data types and compile summary information**




# **PACE stages**

Throughout these project notebooks, I'll see references to the problem-solving framework PACE. The following notebook components are labeled with the respective PACE stage: Plan, Analyze, Construct, and Execute.



## **PACE: Plan**

Considering the questions in my PACE Strategy Document and those below to craft my response:



### **Task 1. Understand the situation**

*   How can I best prepare to understand and organize the provided information?


*Begin by exploring your dataset and consider reviewing the Data Dictionary.*

>Prepare by reading in the data, viewing the data dictionary, and exploring the dataset to identify key variables for the stakeholder.



## **PACE: Analyze**

Considering the questions in my PACE Strategy Document to reflect on the Analyze stage.

In [1]:
# Import packages
import pandas as pd
import numpy as np

Creating a dataframe will help me conduct data manipulation, exploratory data analysis (EDA), and statistical activities.


In [2]:
# Load dataset into dataframe
data = pd.read_csv("data/tiktok_dataset.csv")

### **Task 2b. Understand the data - Inspect the data**

View and inspect summary information about the dataframe by **coding the following:**

1. `data.head(10)`
2. `data.info()`
3. `data.describe()`

*Consider the following questions:*

**Question 1:** When reviewing the first few rows of the dataframe, what do we observe about the data? What does each row represent?

**Question 2:** When reviewing the `data.info()` output, what do we notice about the different variables? Are there any null values? Are all of the variables numeric? Does anything else stand out?

**Question 3:** When reviewing the `data.describe()` output, what do we notice about the distributions of each variable? Are there any questionable values? Does it seem that there are outlier values?

















In [4]:
# Display and examine the first ten rows of the dataframe
data.head(10)


# claim_status    video_id  video_duration_sec  \
0   1        claim  7017666017                  59   
1   2        claim  4014381136                  32   
2   3        claim  9859838091                  31   
3   4        claim  1866847991                  25   
4   5        claim  7105231098                  19   
5   6        claim  8972200955                  35   
6   7        claim  4958886992                  16   
7   8        claim  2270982263                  41   
8   9        claim  5235769692                  50   
9  10        claim  4660861094                  45   

                            video_transcription_text verified_status  \
0  someone shared with me that drone deliveries a...    not verified   
1  someone shared with me that there are more mic...    not verified   
2  someone shared with me that american industria...    not verified   
3  someone shared with me that the metro of st. p...    not verified   
4  someone shared with me that the number of busi...    not verified   
5  someone shared with me that gross domestic pro...    not verified   
6  someone shared with me that elvis presley has ...    not verified   
7  someone shared with me that the best selling s...    not verified   
8  someone shared with me that about half of the ...    not verified   
9  someone shared with me that it would take a 50...        verified   

  author_ban_status  video_view_count  video_like_count  video_share_count  \
0      under review          343296.0           19425.0              241.0   
1            active          140877.0           77355.0            19034.0   
2            active          902185.0           97690.0             2858.0   
3            active          437506.0          239954.0            34812.0   
4            active           56167.0           34987.0             4110.0   
5      under review          336647.0          175546.0            62303.0   
6            active          750345.0          486192.0           193911.0   
7            active          547532.0            1072.0               50.0   
8            active           24819.0           10160.0             1050.0   
9            active          931587.0          171051.0            67739.0   

   video_download_count  video_comment_count  
0                   1.0                  0.0  
1                1161.0                684.0  
2                 833.0                329.0  
3                1234.0                584.0  
4                 547.0                152.0  
5                4293.0               1857.0  
6                8616.0               5446.0  
7                  22.0                 11.0  
8                  53.0                 27.0  
9                4104.0               2540.0

In [5]:
# Get summary info
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19382 entries, 0 to 19381
Data columns (total 12 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   #                         19382 non-null  int64  
 1   claim_status              19084 non-null  object 
 2   video_id                  19382 non-null  int64  
 3   video_duration_sec        19382 non-null  int64  
 4   video_transcription_text  19084 non-null  object 
 5   verified_status           19382 non-null  object 
 6   author_ban_status         19382 non-null  object 
 7   video_view_count          19084 non-null  float64
 8   video_like_count          19084 non-null  float64
 9   video_share_count         19084 non-null  float64
 10  video_download_count      19084 non-null  float64
 11  video_comment_count       19084 non-null  float64
dtypes: float64(5), int64(3), object(4)
memory usage: 1.8+ MB


In [7]:
# Get summary statistics
data.describe().T

count          mean           std           min  \
#                     19382.0  9.691500e+03  5.595246e+03  1.000000e+00   
video_id              19382.0  5.627454e+09  2.536440e+09  1.234959e+09   
video_duration_sec    19382.0  3.242173e+01  1.622997e+01  5.000000e+00   
video_view_count      19084.0  2.547086e+05  3.228933e+05  2.000000e+01   
video_like_count      19084.0  8.430464e+04  1.334205e+05  0.000000e+00   
video_share_count     19084.0  1.673525e+04  3.203617e+04  0.000000e+00   
video_download_count  19084.0  1.049430e+03  2.004300e+03  0.000000e+00   
video_comment_count   19084.0  3.493121e+02  7.996389e+02  0.000000e+00   

                               25%           50%           75%           max  
#                     4.846250e+03  9.691500e+03  1.453675e+04  1.938200e+04  
video_id              3.430417e+09  5.618664e+09  7.843960e+09  9.999873e+09  
video_duration_sec    1.800000e+01  3.200000e+01  4.700000e+01  6.000000e+01  
video_view_count      4.942500e+03  9.954500e+03  5.043270e+05  9.998170e+05  
video_like_count      8.107500e+02  3.403500e+03  1.250200e+05  6.578300e+05  
video_share_count     1.150000e+02  7.170000e+02  1.822200e+04  2.561300e+05  
video_download_count  7.000000e+00  4.600000e+01  1.156250e+03  1.499400e+04  
video_comment_count   1.000000e+00  9.000000e+00  2.920000e+02  9.599000e+03

**Response:**

>**Question 1:** The dataframe contains a collection of categorical, text, and numerical data. Each row represents a distinct TikTok video that presents either a claim or an opinion and the accompanying metadata about that video.

>**Question 2:** The dataframe contains five float64s, three int64s, and four objects. There are 19,382 observations, but some of the variables are missing values, including claim status, the video transcripton, and all of the count variables.

>**Question 3:** Many of the count variables seem to have outliers at the high end of the distribution. They have very large standard deviations and maximum values that are very high compared to their quartile values.

### **Task 2c. Understand the data - Investigate the variables**

In this phase, I will begin to investigate the variables more closely to better understand them.

From the project proposal, the ultimate objective is to use machine learning to classify videos as either claims or opinions. A good first step towards understanding the data might therefore be examining the `claim_status` variable. I will begin by determining how many videos there are for each different claim status.

In [11]:
# What are the different values for claim status and how many of each are in the data?
data['claim_status'].value_counts(dropna=False)


claim      9608
opinion    9476
NaN         298
Name: claim_status, dtype: int64

**Question:** What do we notice about the values shown?

Next, examine the engagement trends associated with each different claim status.

Starting by using Boolean masking to filter the data according to claim status, then calculating the mean and median view counts for each claim status.

In [46]:
# What is the average view count of videos with "claim" status?
data[data.claim_status=='claim'].groupby('claim_status').agg({'video_view_count': ['mean', 'median']})


video_view_count          
                         mean    median
claim_status                           
claim           501029.452748  501555.0

In [45]:
# What is the average view count of videos with "opinion" status?
data[data.claim_status=='opinion'].groupby('claim_status').agg({'video_view_count': ['mean', 'median']})


video_view_count        
                         mean  median
claim_status                         
opinion            4956.43225  4953.0

**Question:** What do we notice about the mean and media within each claim category?

Now, examining trends associated with the ban status of the author.

I will use `groupby()` to calculate how many videos there are for each combination of categories of claim status and author ban status.

In [47]:
# Get counts for each group combination of claim status and author ban status
data.groupby(['claim_status', 'author_ban_status']).agg({'#': 'count'})


#
claim_status author_ban_status      
claim        active             6566
             banned             1439
             under review       1603
opinion      active             8817
             banned              196
             under review        463

**Question:** What do we notice about the number of claims videos with banned authors? Why might this relationship occur?

>There are many more claim videos with banned authors than there are opinion videos with banned authors. This could mean a number of things, including the possibilities that:
>* Claim videos are more strictly policed than opinion videos
>* Authors must comply with a stricter set of rules if they post a claim than if they post an opinion

>Also, it should be noted that there's no way of knowing if claim videos are inherently more likely than opinion videos to result in author bans, or if authors who post claim videos are more likely to post videos that violate terms of service.

>Finally, while we can use this data to draw conclusions about banned/active authors, we cannot draw conclusions about banned videos. There's no way of determining whether a particular video _caused_ the ban, and banned authors could have posted videos that complied with the terms of service.

Continue investigating engagement levels, now focusing on `author_ban_status`.

Calculating the median video share count of each author ban status.

In [34]:
# What's the median video share count of each author ban status?
data.groupby('author_ban_status').agg({'video_share_count': 'median'})

video_share_count
author_ban_status                   
active                         437.0
banned                       14468.0
under review                  9444.0

**Question:** What do we notice about the share count of banned authors, compared to that of active authors? 

> Banned authors have a median share count that's 33 times the median share count of active authors! 

I will use `groupby()` to group the data by `author_ban_status`, then use `agg()` to get the count, mean, and median of each of the following columns:

* `video_view_count`
* `video_like_count`
* `video_share_count`


In [39]:
data.groupby('author_ban_status').agg({'video_view_count': ['count','sum','median'],
                                      'video_like_count': ['count','sum','median'],
                                      'video_share_count': ['count','sum','median']})

video_view_count                         video_like_count  \
                             count           sum    median            count   
author_ban_status                                                             
active                       15383  3.321606e+09    8616.0            15383   
banned                        1635  7.289573e+08  448201.0             1635   
under review                  2066  8.102952e+08  365245.5             2066   

                                          video_share_count               \
                            sum    median             count          sum   
author_ban_status                                                          
active             1.092755e+09    2222.0             15383  217076684.0   
banned             2.501832e+08  105573.0              1635   49048271.0   
under review       2.659315e+08   71204.5              2066   53250524.0   

                            
                    median  
author_ban_status           
active               437.0  
banned             14468.0  
under review        9444.0

**Question:** What do we notice about the number of views, likes, and shares for banned authors compared to active authors?

>A few observations stand out:
* Banned authors and those under review get far more views, likes, and shares than active authors.
* In most groups, the mean is much greater than the median, which indicates that there are some videos with very high engagement counts.

Now, create three new columns to help better understand engagement rates:
* `likes_per_view`: represents the number of likes divided by the number of views for each video
* `comments_per_view`: represents the number of comments divided by the number of views for each video
* `shares_per_view`: represents the number of shares divided by the number of views for each video

In [41]:
# Create a likes_per_view column
data['likes_per_view'] = data['video_like_count']/data['video_view_count']

# Create a comments_per_view column
data['comments_per_view'] = data['video_comment_count']/data['video_view_count']

# Create a shares_per_view column
data['shares_per_view'] = data['video_share_count']/data['video_view_count']

Using `groupby()` to compile the information in each of the three newly created columns for each combination of categories of claim status and author ban status, then use `agg()` to calculate the count, the mean, and the median of each group.

In [42]:
data.groupby(['claim_status', 'author_ban_status']).agg({'likes_per_view': ['count','sum','median'],
                                      'comments_per_view': ['count','sum','median'],
                                      'shares_per_view': ['count','sum','median']})


likes_per_view                         \
                                        count          sum    median   
claim_status author_ban_status                                         
claim        active                      6566  2163.772970  0.326538   
             banned                      1439   496.556528  0.358909   
             under review                1603   525.778643  0.320867   
opinion      active                      8817  1937.478628  0.218330   
             banned                       196    40.546207  0.198483   
             under review                 463   104.820594  0.228051   

                               comments_per_view                      \
                                           count       sum    median   
claim_status author_ban_status                                         
claim        active                         6566  9.144001  0.000776   
             banned                         1439  1.981775  0.000746   
             under review                   1603  2.191442  0.000789   
opinion      active                         8817  4.559119  0.000252   
             banned                          196  0.085135  0.000193   
             under review                    463  0.247962  0.000293   

                               shares_per_view                        
                                         count         sum    median  
claim_status author_ban_status                                        
claim        active                       6566  429.782704  0.049279  
             banned                       1439   97.698624  0.051606  
             under review                 1603  105.370751  0.049967  
opinion      active                       8817  385.554840  0.032405  
             banned                        196    7.944022  0.030728  
             under review                  463   20.590731  0.035027

**Question:**

How does the data for claim videos and opinion videos compare or differ? Consider views, comments, likes, and shares.

>We know that videos by banned authors and those under review tend to get far more views, likes, and shares than videos by non-banned authors. However, *when a video does get viewed*, its engagement rate is less related to author ban status and more related to its claim status.

>Also, we know that claim videos have a higher view rate than opinion videos, but this tells us that claim videos also have a higher rate of likes on average, so they are more favorably received as well. Furthermore, they receive more engagement via comments and shares than opinion videos.

>Note that for claim videos, banned authors have slightly higher likes/view and shares/view rates than active authors or those under review. However, for opinion videos, active authors and those under review both get higher engagement rates than banned authors in all categories.



## **PACE: Construct**

**Note**: The Construct stage does not apply to this workflow. The PACE framework can be adapted to fit the specific requirements of any project.






## **PACE: Execute**

Considering the questions in your PACE Strategy Document and those below to craft my response.

### **What can I summarize for Rosie Mae Bradshaw and the TikTok data team?**

*TikTok's request for a summary is to cover the following points:*

*   What percentage of the data is comprised of claims and what percentage is comprised of opinions?
*   What factors correlate with a video's claim status?
*   What factors correlate with a video's engagement level?


>* Of the 19,382 samples in this dataset, just under 50% are claims&mdash;9,608 of them.  
>* Engagement level is strongly correlated with claim status. This should be a focus of further inquiry.
>* Videos with banned authors have significantly higher engagement than videos with active authors. Videos with authors under review fall between these two categories in terms of engagement levels.